pkg/assets/create/creater.go

```diff
@@ -0,0 +1,232 @@
+package create
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/ghodss/yaml"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+
+	"github.com/openshift/library-go/pkg/assets"
+)
+
+// CreateOptions allow to specify additional create options.
+type CreateOptions struct {
+	// Filters allows to filter which files we will read from disk.
+	// Multiple filters can be specified, in that case only files matching all filters will be returned.
+	Filters []assets.FileInfoPredicate
+
+	// Verbose if true will print out extra messages for debugging
+	Verbose bool
+
+	// StdErr allows to override the standard error output for printing verbose messages.
+	// If not set, os.StdErr is used.
+	StdErr io.Writer
+}
+
+// EnsureManifestsCreated ensures that all resource manifests from the specified directory are created.
+// This function will keep retrying creation until no errors are reported.
+// Pass the context to indicate how much time you are willing to wait until all resources are created.
+func EnsureManifestsCreated(ctx context.Context, manifestDir string, restConfig *rest.Config, options CreateOptions) error {
+	client, dc, err := getClientsFn(restConfig)
+	if err != nil {
+		return err
+	}
+
+	manifests, err := load(manifestDir, options)
+	if err != nil {
+		return err
+	}
+
+	if options.Verbose && options.StdErr == nil {
+		options.StdErr = os.Stderr
+	}
+
+	// Default QPS in client (when not specified) is 5 requests/per second
+	// This specify the interval between "create-all-resources", no need to make this configurable.
+	interval := 200 * time.Millisecond
+
+	// Retry creation until no errors are returned or the timeout is hit.
+	var (
+		lastCreateError      error
+		retryCount           int
+		mapper               meta.RESTMapper
+		needDiscoveryRefresh bool = true
+	)
+	err = wait.PollImmediateUntil(interval, func() (bool, error) {
+		retryCount++
+		// If we get rest mapper error, we need to pull updated discovery info from API server
+		if needDiscoveryRefresh {
+			mapper, err = fetchLatestDiscoveryInfoFn(dc)
+			if err != nil {
+				return false, err
+			}
+		}
+		err, needDiscoveryRefresh = create(manifests, client, mapper, options)
+		if err == nil {
+			// No errors means all resources were created successfully, reset lastCreateError and exit the loop.
+			lastCreateError = nil
```

Comment:
why not assign directly, `lastCreateError, needDiscoveryRefresh = create(...)` ?